# Introduction
This assignment will test how well you're able to perform various data science-related tasks.

Each Problem Group below will center around a particular dataset that you have worked with before.

To ensure you receive full credit for a question, make sure you demonstrate the appropriate pandas, altair, or other commands as requested in the provided code blocks.

You may find that some questions require multiple steps to fully answer. Others require some mental arithmetic in addition to pandas commands. Use your best judgment.

## Submission
Each problem group asks a series of questions. This assignment consists of two submissions:

1. After completing the questions below, open the Module 01 Assessment Quiz in Canvas and enter your answers to these questions there.

2. After completing and submitting the quiz, save this Colab notebook as a GitHub Gist (You'll need to create a GitHub account for this), by selecting `Save a copy as a GitHub Gist` from the `File` menu above.

    In Canvas, open the Module 01 Assessment GitHub Gist assignment and paste the GitHub Gist URL for this notebook. Then submit that assignment.

## Problem Group 1

For the questions in this group, you'll work with the Netflix Movies Dataset found at this url: [https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/netflix_titles.csv](https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/netflix_titles.csv)


### Question 1
Load the dataset into a Pandas data frame and determine what data type is used to store the `release_year` feature.

### Question 2
Filter your dataset so it contains only `TV Shows`. How many of those TV Shows were rated `TV-Y7`?

### Question 3
Further filter your dataset so it only contains TV Shows released between the years 2000 and 2009 inclusive. How many of *those* shows were rated `TV-Y7`?

## Problem Group 2

For the questions in this group, you'll work with the Cereal Dataset found at this url: [https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/cereal.csv](https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/cereal.csv)


### Question 4
After importing the dataset into a pandas data frame, determine the median amount of `protein` in cereal brands manufactured by Kelloggs. (`mfr` code "K")

### Question 5
In order to comply with new government regulations, all cereals must now come with a "Healthiness" rating. This rating is calculated based on this formula:

    healthiness = (protein + fiber) / sugar

Create a new `healthiness` column populated with values based on the above formula.

Then, determine the median healthiness value for only General Mills cereals (`mfr` = "G"), rounded to two decimal places.

## Problem Group 3

For the questions in this group, you'll work with the Titanic Dataset found at this url: [https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/titanic.csv](https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/titanic.csv)

### Question 6

After loading the dataset into a pandas DataFrame, create a new column called `NameGroup` that contains the first letter of the passenger's surname in lower case.

Note that in the dataset, passenger's names are provided in the `Name` column and are listed as:

    Surname, Given names

For example, if a passenger's `Name` is `Braund, Mr. Owen Harris`, the `NameGroup` column should contain the value `b`.

Then count how many passengers have a `NameGroup` value of `k`.

# Task
Answer all questions using the "netflix_titles.csv", "cereal.csv", and "titanic.csv" datasets.

## Question 1

### Subtask:
Load the Netflix dataset, determine and display the data type of the `release_year` column.


**Reasoning**:
The first step is to load the dataset and display the data type of the 'release_year' column.



In [9]:
import pandas as pd

df_netflix = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/netflix_titles.csv')
display(df_netflix['release_year'].dtype)

dtype('int64')

## Problem group 1: netflix dataset

### Subtask:
Filter the Netflix dataset for TV Shows and count how many are rated `TV-Y7`.


**Reasoning**:
Filter the DataFrame for TV Shows and count the number of rows with a 'TV-Y7' rating, then display the count.



In [10]:
tv_shows_df = df_netflix[df_netflix['type'] == 'TV Show']
tv_y7_count = tv_shows_df[tv_shows_df['rating'] == 'TV-Y7'].shape[0]
display(tv_y7_count)

100

## Problem group 1: netflix dataset

### Subtask:
Filter the TV Shows dataset to include only those released between 2000 and 2009 inclusive, and count how many of these are rated `TV-Y7`.


**Reasoning**:
Filter the tv_shows_df to include only TV shows released between 2000 and 2009 (inclusive) and then count the number of those shows rated 'TV-Y7'.



In [11]:
tv_shows_2000s_df = tv_shows_df[(tv_shows_df['release_year'] >= 2000) & (tv_shows_df['release_year'] <= 2009)]
tv_y7_2000s_count = tv_shows_2000s_df[tv_shows_2000s_df['rating'] == 'TV-Y7'].shape[0]
display(tv_y7_2000s_count)

4

## Problem group 2: cereal dataset

### Subtask:
Load the Cereal dataset, filter for Kellogg's cereals (`mfr` code "K"), and determine the median `protein` amount for this subset.


**Reasoning**:
Load the cereal dataset, filter for Kellogg's cereals, calculate the median protein, and display the result.



In [12]:
df_cereal = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/cereal.csv')
kelloggs_cereal_df = df_cereal[df_cereal['mfr'] == 'K']
median_protein_kelloggs = kelloggs_cereal_df['protein'].median()
display(median_protein_kelloggs)

3.0

## Problem group 2: cereal dataset

### Subtask:
Create a new `healthiness` column in the Cereal dataset using the formula `(protein + fiber) / sugar`. Filter for General Mills cereals (`mfr` code "G"), calculate the median `healthiness` for this subset, and round the result to two decimal places.


**Reasoning**:
Calculate the healthiness column, filter for General Mills cereals, calculate the median healthiness, round the result, and display it.



In [19]:
df_cereal['healthiness'] = (df_cereal['protein'] + df_cereal['fiber']) / df_cereal['sugars']
general_mills_cereal_df = df_cereal[df_cereal['mfr'] == 'G']
median_healthiness_general_mills = general_mills_cereal_df['healthiness'].median()
rounded_median_healthiness = round(median_healthiness_general_mills, 2)
display(rounded_median_healthiness)

0.47

## Problem group 3: titanic dataset

### Subtask:
Load the Titanic dataset, create a new column `NameGroup` containing the first letter of the surname in lowercase. Count and display the number of passengers with a `NameGroup` value of 'k'.

**Reasoning**:
Load the Titanic dataset, create the `NameGroup` column by extracting the first letter of the surname, filter for 'k' in `NameGroup`, count the filtered rows, and display the count.

In [18]:
df_titanic = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/titanic.csv')

def get_surname_initial(name):
    if ',' in name:
        surname = name.split(',')[0]
        return surname[0].lower()
    return ''

df_titanic['NameGroup'] = df_titanic['Name'].apply(get_surname_initial)

k_namegroup_count = df_titanic[df_titanic['NameGroup'] == 'k'].shape[0]

display(k_namegroup_count)

28

**Reasoning**:
The previous attempt failed because the column 'sugar' was not found. I need to check the available columns in `df_cereal` to make sure the column name is correct, then regenerate the code block.



In [14]:
print(df_cereal.columns)

Index(['name', 'mfr', 'type', 'calories', 'protein', 'fat', 'sodium', 'fiber',
       'carbo', 'sugars', 'potass', 'vitamins', 'shelf', 'weight', 'cups',
       'rating'],
      dtype='object')


**Reasoning**:
The column name for sugar is 'sugars', not 'sugar'. I will regenerate the code to calculate healthiness, filter, calculate the median, round, and display the result using the correct column name.



In [15]:
df_cereal['healthiness'] = (df_cereal['protein'] + df_cereal['fiber']) / df_cereal['sugars']
general_mills_cereal_df = df_cereal[df_cereal['mfr'] == 'G']
median_healthiness_general_mills = general_mills_cereal_df['healthiness'].median()
rounded_median_healthiness = round(median_healthiness_general_mills, 2)
display(rounded_median_healthiness)

0.47

## Problem group 3: titanic dataset

### Subtask:
Load the Titanic dataset, create a new column `NameGroup` containing the first letter of the surname in lowercase. Count and display the number of passengers with a `NameGroup` value of 'k'.


**Reasoning**:
Load the Titanic dataset, create the `NameGroup` column by extracting the first letter of the surname, filter for 'k' in `NameGroup`, count the filtered rows, and display the count.



In [16]:
df_titanic = pd.read_csv('https://raw.githubusercontent.com/byui-cse/cse450-course/master/data/titanic.csv')

def get_surname_initial(name):
    if ',' in name:
        surname = name.split(',')[0]
        return surname[0].lower()
    return ''

df_titanic['NameGroup'] = df_titanic['Name'].apply(get_surname_initial)

k_namegroup_count = df_titanic[df_titanic['NameGroup'] == 'k'].shape[0]

display(k_namegroup_count)

28

## Summary:

### Q&A
*   How many TV Shows in the Netflix dataset are rated TV-Y7?
    There are 100 TV Shows in the Netflix dataset that are rated TV-Y7.
*   How many TV Shows in the Netflix dataset released between 2000 and 2009 are rated TV-Y7?
    There are 4 TV Shows in the Netflix dataset released between 2000 and 2009 that are rated TV-Y7.
*   What is the median protein amount for Kellogg's cereals in the Cereal dataset?
    The median protein amount for Kellogg's cereals is 3.0.
*   What is the median healthiness score for General Mills cereals in the Cereal dataset, rounded to two decimal places?
    The median healthiness score for General Mills cereals is 0.47.
*   How many passengers in the Titanic dataset have a surname starting with the letter 'K'?
    There are 28 passengers in the Titanic dataset whose surname starts with the letter 'K'.

### Data Analysis Key Findings
*   The `release_year` column in the Netflix dataset has a data type of `int64`.
*   There are 100 TV Shows rated `TV-Y7` in the Netflix dataset.
*   Out of the TV Shows released between 2000 and 2009, 4 are rated `TV-Y7`.
*   The median protein amount for Kellogg's cereals is 3.0.
*   The median healthiness score (calculated as `(protein + fiber) / sugars`) for General Mills cereals is 0.47, when rounded to two decimal places.
*   There are 28 passengers in the Titanic dataset whose surname begins with the letter 'K'.

### Insights or Next Steps
*   The relatively low number of TV-Y7 rated shows from 2000-2009 compared to the total suggests a potential shift in content rating distribution over time or a focus on different demographics during that period.
*   Analyzing the distribution of healthiness scores across different cereal manufacturers could provide insights into industry trends regarding nutritional content.
